src/Microsoft.ML.GenAI.Core/Pipeline/CausalLMPipeline.cs

```diff
@@ -7,128 +7,266 @@
+    ) Generate(
+        Tensor inputIds, // input token ids [batch_size, sequence_length]
+        Tensor attentionMask, // attention mask [batch_size, sequence_length]
+        int[][] stopTokenSequence,
+        float temperature = 0.7f,
+        float topP = 0.9f,
+        int maxLen = 128)
+    {
+        using var scope = NewDisposeScope();
+        Tensor? logits = null;
+        foreach (var (token, _logits) in this.GenerateStreaming(inputIds, attentionMask, stopTokenSequence, temperature, topP, maxLen))
+        {
+            inputIds = torch.cat([inputIds, token.unsqueeze(1)], dim: -1).MoveToOtherDisposeScope(inputIds);
+            if (logits is null)
             {
-                // return entire inputIds and logits
-                return (inputIds.MoveToOuterDisposeScope(), logits!.MoveToOuterDisposeScope());
+                logits = _logits;
             }
             else
             {
-                // return [batch_size, promptLength..] and [batch_size, promptLength.., vocab_size]
-                return (inputIds[.., promptLength..].MoveToOuterDisposeScope(), logits![.., promptLength..].MoveToOuterDisposeScope());
+                logits = torch.cat([logits, _logits], dim: -1).MoveToOtherDisposeScope(inputIds);
             }
         }
+
+        return (inputIds, logits ?? throw new InvalidOperationException("Logits is null"));
+    }
+
+    public virtual string Generate(
+        string prompt,
+        int maxLen = 128,
+        float temperature = 0.7f,
+        float topP = 0.9f,
+        string[]? stopSequences = null)
+    {
+        var chunks = new List<string>();
+
+        foreach (var chunk in this.GenerateStreaming(prompt, maxLen, temperature, topP, stopSequences))
+        {
+            chunks.Add(chunk);
+        }
+
+        return string.Join(string.Empty, chunks);
+    }
+
+
+    public virtual IEnumerable<string> GenerateStreaming(
+        string prompt,
+        int maxLen = 128,
+        float temperature = 0.7F,
+        float topP = 0.9F,
+        string[]? stopSequences = Defaults.StopSequence)
+    {
+        using var newScope = NewDisposeScope();
+        var inputIds = this.Tokenizer.EncodeToIds(prompt);
+        var inputTensor = torch.tensor(inputIds.ToArray(), dtype: ScalarType.Int64, device: this.Device).unsqueeze(0);
+        var attentionMask = torch.ones_like(inputTensor, device: this.Device);
+        // set up stop token ids
+        // stop token ids: [[eosId], [stopSequence1], [stopSequence2], ...]
+        // when causal language model generates tokens, it will stop when it generates any token in stopSequences
+        List<int[]> stopTokenIds = [[]];
+        if (stopSequences != null)
+        {
+            stopTokenIds.AddRange(stopSequences.Select(x =>
+            {
+                var tokens = this.Tokenizer.EncodeToTokens(x, out var _, false, false);
+
+                return tokens
+                .Where(t => t.Offset != (0, 0))
+                .Select(t => t.Id)
```

Comment:
what this for?

Reply:
To detect the `_` token automatically added by tokenzier

Reply:
ok, thanks. it is worth a comment. 

Also, EncodeToToken is very expensive comparing to EncodeToIds. If you are interested to optimize I can suggest some code to check for `_ ` token.

Reply:
Yes that would be very helpful